spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/security/IcebergHiveConnectorDelegationTokenProvider.java

```diff
@@ -0,0 +1,240 @@
+            })
+        .map(Tuple2::_1)
+        .collect(Collectors.toSet());
+  }
+
+  @Override
+  public boolean delegationTokensRequired(SparkConf sparkConf, Configuration hadoopConf) {
+    return !getRequireTokenCatalogs(sparkConf).isEmpty();
+  }
+
+  private Set<String> getRequireTokenCatalogs(SparkConf sparkConf) {
+    return extractIcebergCatalogNames(sparkConf).stream()
+        .filter(catalog -> checkDelegationTokensRequired(sparkConf, catalog))
+        .collect(Collectors.toSet());
+  }
+
+  private boolean checkDelegationTokensRequired(SparkConf sparkConf, String catalogName) {
+    String metastoreUri = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + URI_KEY, "");
+    String principal = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + PRINCIPAL_KEY, "");
+    boolean isHiveType =
+        CATALOG_TYPE.equalsIgnoreCase(
+            sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + TYPE_KEY, ""));
+    if (metastoreUri.isEmpty()
+        || principal.isEmpty()
+        || !isHiveType
+        || !UserGroupInformation.isSecurityEnabled()) {
+      return false;
+    }
+
+    try {
+      Credentials credentials = UserGroupInformation.getCurrentUser().getCredentials();
+      Token<?> currentToken = credentials.getToken(new Text(metastoreUri));
+      return currentToken == null;
+    } catch (IOException ex) {
+      LOG.error(
+          "Failed to get current user credentials for catalog {}: {}",
+          catalogName,
+          ex.getMessage(),
+          ex);
+      throw new RuntimeException(ex);
+    }
+  }
+
+  IMetaStoreClient createHmsClient(HiveConf conf) throws Exception {
+    return new HiveMetaStoreClient(conf, null, false);
+  }
+
+  @Override
+  public Option<Object> obtainDelegationTokens(
+      Configuration hadoopConf, SparkConf sparkConf, Credentials creds) {
+    Map<String, IMetaStoreClient> hmsClientCache = Maps.newHashMap();
+    Set<String> requireTokenCatalogs = getRequireTokenCatalogs(sparkConf);
+    LOG.debug("Require token Hive catalogs: {}", requireTokenCatalogs);
+    try {
+      for (String catalogName : requireTokenCatalogs) {
+        Optional<HiveConf> hiveConfOpt = buildHiveConf(sparkConf, hadoopConf, catalogName);
+        if (!hiveConfOpt.isPresent()) {
+          continue;
+        }
+
+        HiveConf remoteHmsConf = hiveConfOpt.get();
+        String metastoreUri = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + URI_KEY);
+        String principal = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + PRINCIPAL_KEY);
+        try {
+          UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+          LOG.debug(
+              "Getting Hive delegation token for {} against {} at {}",
+              currentUser.getUserName(),
+              principal,
+              metastoreUri);
+          doAsRealUser(
+              () -> {
+                IMetaStoreClient hmsClient = hmsClientCache.get(metastoreUri);
```

Comment:
When we use any caching then we must make sure that when the client is corrupted (constantly throws exception) then the client must be evicted and newly created, otherwise I don't see the gain to have a cache. Do we really need a cache here? I mean, maybe we can space 2-3 seconds within a 4 hours period.